src/test/store/js-tracer.test.js

```diff
@@ -0,0 +1,170 @@
+        'https://mozilla.org (0:20)',
+        'Interpreter (1:19)',
+        'IonMonkey (2:18)',
+      ]);
+    });
+
+    it('can generate sibling timing', function() {
+      expect(
+        getHumanReadableJsTracerTiming({
+          selfTime: false,
+          events: [
+            ['https://mozilla.org', 0, 20],
+            ['Interpreter', 1, 5],
+            ['Interpreter', 2, 4],
+            ['IonMonkey', 5, 19],
+            ['IonMonkey', 6, 18],
+          ],
+        })
+      ).toEqual([
+        'https://mozilla.org (0:20)',
+        'Interpreter (1:5) | IonMonkey (5:19)',
+        'Interpreter (2:4) | IonMonkey (6:18)',
+      ]);
+    });
+  });
+  describe('self time', function() {
+    it('has empty JS tracer timing if no events are in the js tracer table', function() {
+      expect(
+        getHumanReadableJsTracerTiming({ selfTime: true, events: [] })
+      ).toEqual([]);
+    });
+
+    it('can generate some simple nested timing', function() {
+      expect(
+        getHumanReadableJsTracerTiming({
+          selfTime: true,
+          events: [
+            ['https://mozilla.org', 0, 20],
+            ['Interpreter', 1, 19],
+            ['IonMonkey', 2, 18],
+          ],
+        })
+      ).toEqual([
+        'Interpreter (1:2) | Interpreter (18:19)',
+        'IonMonkey (2:18)',
+        'https://mozilla.org (0:1) | https://mozilla.org (19:20)',
+      ]);
+    });
+
+    it('can "drain off" prefixes the first if branch of the algorithm', function() {
+      // This test is checking a specific if branch of the timing function:
+      //
+      //    xxxxxxxxxxxxxxxx[================]
+      //    xxxxxxxx[======]     [current]
+      //    [prefix]
+      expect(
+        getHumanReadableJsTracerTiming({
+          selfTime: true,
+          events: [
+            // Make the formatting "prettier".
+            ['A', 0, 2],
+            ['B', 0, 2],
+            ['C', 0, 1],
+            ['A', 2, 3],
+          ],
+        })
+      ).toEqual([
+        // Make the formatting "prettier".
+        'A (2:3)',
+        'B (1:2)',
+        'C (0:1)',
+      ]);
+    });
+
+    it('can split a prefix stack if the ending time matches', function() {
+      // This test is checking a specific if branch of the timing function:
+      //
+      //   [prefix]xxxxxxxxx
+      //           [current]
+      expect(
+        getHumanReadableJsTracerTiming({
+          selfTime: true,
+          events: [
+            // Make the formatting "prettier".
+            ['A', 0, 2],
+            ['B', 1, 2],
+          ],
+        })
+      ).toEqual([
+        // Make the formatting "prettier".
+        'A (0:1)',
+        'B (1:2)',
+      ]);
+    });
+  });
+});
+
+/**
+ * Create JS tracing information from a list of events, and then compute a human
+ * readable representation of the timing.
+ *
+ * "EventName1 (StartTime:EndTime) | EventName2 (StartTime:EndTime)"
+ */
+function getHumanReadableJsTracerTiming({
```

Comment:
I'd split the function in 2 parts:
1. part 1: a `setup` function that generates a profile and dispatch
2. part 2: the part with  `ensureExists(...).map(...)`

And keep the call to `getExpensiveJsTracerTiming` in the test itself.
I think this would make it easier to follow the flow of the tests.

Reply:
I'm trying to understand how this would make the test flow better, as we'll need to pass around a few transient variables. I almost wonder if I should take this selector out, since we're only using it in tests.

Reply:
yeah I also advise about removing `getExpensiveJsTracerTiming` from our code elsewhere :)

About this specific function, the problem for me is that it does several things: doing a setup, and retrieving data from the store. But the name of the function itself doesn't say that. I find that it makes the tests a bit confusing to read, as it's difficult to get an overview about what happens without looking at the details of this function.

This is an optional comment though, if you still feel that it's not useful I'm fine.

Reply:
Fixed by removing the selector.